clientapi/routing/register.go

```diff
@@ -83,23 +83,17 @@ func (d sessionsDict) GetCompletedStages(sessionID string) []authtypes.LoginType
 	return make([]authtypes.LoginType, 0)
 }
 
-// AddCompletedStage records that a session has completed an auth stage.
-func (d *sessionsDict) AddCompletedStage(sessionID string, stage authtypes.LoginType) {
-	// Return if the stage is already present
-	for _, completedStage := range d.GetCompletedStages(sessionID) {
-		if completedStage == stage {
-			return
-		}
-	}
-	d.sessions[sessionID] = append(d.GetCompletedStages(sessionID), stage)
-}
-
 func newSessionsDict() *sessionsDict {
 	return &sessionsDict{
 		sessions: make(map[string][]authtypes.LoginType),
 	}
 }
 
+// AddCompletedSessionStage records that a session has completed an auth stage.
+func AddCompletedSessionStage(sessionID string, stage authtypes.LoginType) {
+	sessions.sessions[sessionID] = append(sessions.GetCompletedStages(sessionID), stage)
```

Comment:
I think we are just one (or two) things away from finishing this PR.

First thing is that `AddCompletedSessionStage` should, like `func (d *sessionsDict) AddCompletedStage`, first check if there's already such a stage in the list, and if so, don't add a duplicate one.

Second thing is to fix https://github.com/matrix-org/dendrite/issues/780 by adding a mutex to `sessions`. Of course, if you don't have enough time, feel free to leave the second thing out and we can still get it done in another PR after merging this one :)

Reply:
Added the duplication check, but this causes tests to fail.